### PR TITLE
Event notifications

### PR DESCRIPTION
**Pull request by raonismaneoto**
**Feature/Bug description:**
Send notifications to the user who chose to follow the event when it gets updated or deleted.

**Solution:**
I've included another attribute in the event's model called followers. The user who clicks on follow event is added in this attribute. The worker iterates over this attribute sending notification for each follower when the event gets updated or deleted.

screenshot from 2019-02-12 10-18-50

**TODO/FIXME:** n/a

### CODE REVIEW

File: backend/test/event_handlers_tests/event_handler_test.py [hunk 1]
```diff
@@ -154,11 +169,12 @@ def test_patch(self, verify_token):
...
                                     [{"op": "replace", "path": "/local",
-                                      "value": "New Local"}]
-                                    )
+                                      "value": "New Local"}],
+                                    headers={'institution-authorization': self.institution.key.urlsafe()})
 
+    @patch('handlers.event_handler.enqueue_task')
     @patch('util.login_service.verify_token', return_value={'email': 'user@gmail.com'})
-    def test_pacth_datetime(self, verify_token):
+    def test_pacth_datetime(self, verify_token, mock_method):
```
Comment (ruanres) [inline backend/test/event_handlers_tests/event_handler_test.py]: use the name of the method that you are mocking

File: frontend/event/eventDetailsDirective.js [hunk 2]
```diff
@@ -148,6 +148,28 @@
             return new Date(isoTime).getHours();
         };
 
+        eventCtrl.isFollower = () => {
+            return eventCtrl.event && _.includes(eventCtrl.event.followers, eventCtrl.user.key);
```
Comment (ruanres) [inline frontend/event/eventDetailsDirective.js]: there's no need to use lodash 
`eventCtrl.event.followers.includes(eventCtrl.user.key)`

File: frontend/event/event.js [hunk 3]
```diff
@@ -25,3 +25,13 @@ Event.prototype.convertDate = function(){
     }
 };
 
+Event.prototype.addFollower = function (userKey) {
+    this.followers.push(userKey);
+};
+
+Event.prototype.removeFollower = function (userKey) {
+    _.remove(this.followers, followerKey => {
```
Comment (ruanres) [inline frontend/event/event.js]: you can use filter

File: frontend/event/eventService.js [hunk 4]
```diff
@@ -39,5 +39,13 @@
         service.getMonths = function getMonths() {
             return HttpService.get('app/utils/months.json');
         };
+
+        service.addFollower = function (event) {
```
Comment (ruanres) [inline frontend/event/eventService.js]: eventKey

File: frontend/event/eventService.js [hunk 5]
```diff
@@ -39,5 +39,13 @@
...
+            return HttpService.post(`/api/events/${event}/followers`);
+        };
+
+        service.removeFollower = function (event) {
```
Comment (ruanres) [inline frontend/event/eventService.js]: same here

File: frontend/event/event_details.html [hunk 6]
```diff
@@ -46,17 +46,29 @@ <h3 md-colors="{color:'default-teal-400'}" ng-if="!eventDetailsCtrl.endInOtherMo
                 </md-button>
               </md-menu-item>
               <md-menu-item id="hide-share">
-                <md-button ng-click="eventDetailsCtrl.share($event, eventDetailsCtrl.event)"">
+                <md-button ng-click="eventDetailsCtrl.share($event, eventDetailsCtrl.event)">
```
Comment (ruanres) [inline frontend/event/event_details.html]: there is no need to pass the event as a parameter once the controller already have its reference

File: frontend/event/event_details.html [hunk 7]
```diff
@@ -46,17 +46,29 @@ <h3 md-colors="{color:'default-teal-400'}" ng-if="!eventDetailsCtrl.endInOtherMo
...
                   <md-icon>share</md-icon>
                   Compartilhar evento
                 </md-button>
               </md-menu-item>
-              <md-menu-item>
+              <md-menu-item ng-if="eventDetailsCtrl.canChange(eventDetailsCtrl.event)">
```
Comment (ruanres) [inline frontend/event/event_details.html]: same here

File: frontend/post/postDetailsDirective.js [hunk 8]
```diff
@@ -275,6 +275,35 @@
             return Utils.isLargerThanTheScreen(postDetailsCtrl.post.title) ? 'break' : 'no-break';
         };
 
+        postDetailsCtrl.isSharedEvent = () => {
+            return postDetailsCtrl.post.shared_event; 
+        };
+
+        postDetailsCtrl.isFollowingEvent = () => {
+            const eventFollowers = (postDetailsCtrl.post.shared_event && postDetailsCtrl.post.shared_event.followers) || [];
+            return _.includes(eventFollowers, postDetailsCtrl.user.key);
```
Comment (ruanres) [inline frontend/post/postDetailsDirective.js]: no need to use lodash

File: frontend/post/postDetailsDirective.js [hunk 9]
```diff
@@ -275,6 +275,35 @@
...
+                postDetailsCtrl.post.shared_event.followers.push(postDetailsCtrl.user.key);
+                MessageService.showToast('Você receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
+            });
+        };
+
+        postDetailsCtrl.unFollowEvent = () => {
+            EventService.removeFollower(postDetailsCtrl.post.shared_event.key).then(() => {
+                _.remove(postDetailsCtrl.post.shared_event.followers, follower => {
```
Comment (ruanres) [inline frontend/post/postDetailsDirective.js]: filter

File: frontend/test/specs/event/eventDetailsControllerSpec.js [hunk 10]
```diff
@@ -250,4 +251,58 @@
...
+                return q.when();
+            });
+            spyOn(messageService, 'showToast');
+            eventCtrl.event = new Event({ key: 'aopskdopas-OKAPODKAOP', followers: [] });
+            spyOn(eventCtrl.event, 'addFollower').and.callThrough();
+            
+            eventCtrl.addFollower();
+            scope.$apply();
+
+            expect(eventService.addFollower).toHaveBeenCalled();
```
Comment (ruanres) [inline frontend/test/specs/event/eventDetailsControllerSpec.js]: check if it was called with the event key

File: frontend/test/specs/event/eventDetailsControllerSpec.js [hunk 11]
```diff
@@ -250,4 +251,58 @@
...
+                return q.when();
+            });
+            spyOn(messageService, 'showToast');
+            eventCtrl.event = new Event({ key: 'aopskdopas-OKAPODKAOP', followers: [] });
+            spyOn(eventCtrl.event, 'removeFollower').and.callThrough();
+
+            eventCtrl.removeFollower();
+            scope.$apply();
+
+            expect(eventService.removeFollower).toHaveBeenCalled();
```
Comment (ruanres) [inline frontend/test/specs/event/eventDetailsControllerSpec.js]: same here

File: frontend/event/eventDetailsDirective.js [hunk 12]
```diff
@@ -148,6 +148,28 @@
...
+        };
+
+        eventCtrl.addFollower = () => {
+            EventService.addFollower(eventCtrl.event.key).then(() => {
+                eventCtrl.event.addFollower(eventCtrl.user.key);
+                MessageService.showToast('Você receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
```
Comment (JuliePessoa) [inline frontend/event/eventDetailsDirective.js]: console?

File: frontend/event/eventDetailsDirective.js [hunk 13]
```diff
@@ -148,6 +148,28 @@
             return new Date(isoTime).getHours();
         };
 
+        eventCtrl.isFollower = () => {
+            return eventCtrl.event && _.includes(eventCtrl.event.followers, eventCtrl.user.key);
+        };
+
+        eventCtrl.addFollower = () => {
```
Comment (JuliePessoa) [inline frontend/event/eventDetailsDirective.js]: jsdoc

File: frontend/event/eventDetailsDirective.js [hunk 14]
```diff
@@ -148,6 +148,28 @@
...
+            EventService.addFollower(eventCtrl.event.key).then(() => {
+                eventCtrl.event.addFollower(eventCtrl.user.key);
+                MessageService.showToast('Você receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
+            });
+        };
+
+        eventCtrl.removeFollower = () => {
```
Comment (JuliePessoa) [inline frontend/event/eventDetailsDirective.js]: jsdoc

File: frontend/event/eventDetailsDirective.js [hunk 15]
```diff
@@ -148,6 +148,28 @@
...
+            EventService.removeFollower(eventCtrl.event.key).then(() => {
+                eventCtrl.event.removeFollower(eventCtrl.user.key);
+                MessageService.showToast('Você não receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
```
Comment (JuliePessoa) [inline frontend/event/eventDetailsDirective.js]: console?

File: frontend/event/eventService.js (hunk 4 above)
Comment (JuliePessoa) [inline frontend/event/eventService.js]: jsdoc

File: frontend/event/eventDetailsDirective.js [hunk 16]
```diff
@@ -51,10 +51,10 @@
             }
         };
 
-        eventCtrl.canChange = function canChange(event) {
-            if(event) {
-                const hasInstitutionPermission = eventCtrl.user.hasPermission('remove_posts', event.institution_key);
-                const hasEventPermission = eventCtrl.user.hasPermission('remove_post', event.key);
+        eventCtrl.canChange = function canChange() {
```
Comment (JuliePessoa) [inline frontend/event/eventDetailsDirective.js]: jsdoc

File: frontend/post/postDetailsDirective.js [hunk 17]
```diff
@@ -275,6 +275,33 @@
             return Utils.isLargerThanTheScreen(postDetailsCtrl.post.title) ? 'break' : 'no-break';
         };
 
+        postDetailsCtrl.isSharedEvent = () => {
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: jsdoc all functions

File: frontend/post/postDetailsDirective.js [hunk 18]
```diff
@@ -275,6 +275,33 @@
...
+            const eventFollowers = (postDetailsCtrl.post.shared_event && postDetailsCtrl.post.shared_event.followers) || [];
+            return eventFollowers.includes(postDetailsCtrl.user.key);
+        };
+
+        postDetailsCtrl.followEvent = () => {
+            EventService.addFollower(postDetailsCtrl.post.shared_event.key).then(() => {
+                postDetailsCtrl.post.shared_event.followers.push(postDetailsCtrl.user.key);
+                MessageService.showToast('Você receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: console?

File: frontend/post/postDetailsDirective.js [hunk 19]
```diff
@@ -275,6 +275,33 @@
...
+            });
+        };
+
+        postDetailsCtrl.unFollowEvent = () => {
+            EventService.removeFollower(postDetailsCtrl.post.shared_event.key).then(() => {
+                postDetailsCtrl.post.shared_event.followers = postDetailsCtrl.post.shared_event.followers.filter(follower => follower !== postDetailsCtrl.user.key);
+                MessageService.showToast('Você não receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: console?

File: frontend/post/postDetailsDirective.js [hunk 20]
```diff
@@ -275,6 +275,33 @@
...
+        };
+
+        postDetailsCtrl.isFollowingEvent = () => {
+            const eventFollowers = (postDetailsCtrl.post.shared_event && postDetailsCtrl.post.shared_event.followers) || [];
+            return eventFollowers.includes(postDetailsCtrl.user.key);
+        };
+
+        postDetailsCtrl.followEvent = () => {
+            EventService.addFollower(postDetailsCtrl.post.shared_event.key).then(() => {
+                postDetailsCtrl.post.shared_event.followers.push(postDetailsCtrl.user.key);
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: use event prototype function to add follower

File: frontend/post/postDetailsDirective.js [hunk 21]
```diff
@@ -275,6 +275,33 @@
...
+                MessageService.showToast('Você receberá as atualizações desse evento.');
+            }).catch((error) => {
+                console.error(error);
+            });
+        };
+
+        postDetailsCtrl.unFollowEvent = () => {
+            EventService.removeFollower(postDetailsCtrl.post.shared_event.key).then(() => {
+                postDetailsCtrl.post.shared_event.followers = postDetailsCtrl.post.shared_event.followers.filter(follower => follower !== postDetailsCtrl.user.key);
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: use event prototype function to remove follower

File: frontend/test/specs/event/eventDetailsControllerSpec.js [hunk 22]
```diff
@@ -250,4 +252,58 @@
...
+            expect(eventCtrl.isFollower()).toEqual(true);
+        });
+
+        it('should return false when the user is not following the event', () => {
+            expect(eventCtrl.isFollower()).toEqual(false);
+        });
+    });
+
+    describe('addFollower()', () => {
+        it('should call addFollower', () => {
```
Comment (JuliePessoa) [inline frontend/test/specs/event/eventDetailsControllerSpec.js]: test on service error, user is not added

File: frontend/test/specs/event/eventDetailsControllerSpec.js [hunk 23]
```diff
@@ -250,4 +252,58 @@
...
+            scope.$apply();
+
+            expect(eventService.addFollower).toHaveBeenCalledWith(eventCtrl.event.key);
+            expect(messageService.showToast).toHaveBeenCalled();
+            expect(eventCtrl.event.addFollower).toHaveBeenCalled();
+            expect(eventCtrl.event.followers).toEqual([user.key]);
+        });
+    });
+
+    describe('removeFollower()', () => {
```
Comment (JuliePessoa) [inline frontend/test/specs/event/eventDetailsControllerSpec.js]: test on service error, user is not removed

File: frontend/test/specs/event/eventServiceSpec.js [hunk 24]
```diff
@@ -133,5 +133,25 @@
...
                 expect($http.patch).toHaveBeenCalledWith(EVENT_URI + '/' + event.key, patch);
             });
+
+            describe('addFollower', () => {
+                it('should call post', () => {
+                    spyOn($http, 'post');
+
+                    service.addFollower();
+
+                    expect($http.post).toHaveBeenCalled();
```
Comment (JuliePessoa) [inline frontend/test/specs/event/eventServiceSpec.js]: verify post parameters

File: frontend/test/specs/event/eventServiceSpec.js [hunk 25]
```diff
@@ -133,5 +133,25 @@
...
+                });
+            });
+
+            describe('removeFollower', () => {
+                it('should call delete', () => {
+                    spyOn($http, 'delete');
+
+                    service.removeFollower();
+
+                    expect($http.delete).toHaveBeenCalled();
```
Comment (JuliePessoa) [inline frontend/test/specs/event/eventServiceSpec.js]: same

File: frontend/test/specs/event/eventSpec.js [hunk 26]
```diff
@@ -54,4 +55,33 @@
             expect(event.end_time).toEqual(end_time_conv);
         });
     });
+
+    describe('addFollower()', () => {
```
Comment (JuliePessoa) [inline frontend/test/specs/event/eventSpec.js]: add before each for this: `const event = new Event(eventData, institution_key); expect(event.followers).toEqual([]);`

File: frontend/test/specs/post/postDetailsControllerSpec.js [hunk 27]
```diff
@@ -613,4 +617,71 @@
             expect(returnedValue).toBeFalsy();
         });
     });
+
+    describe('isSharedEvent()', () => {
+        it('should be truthy when the post is a shared_event', () => {
+            postDetailsCtrl.post = new Post({shared_event : {}});
```
Comment (JuliePessoa) [inline frontend/test/specs/post/postDetailsControllerSpec.js]: it should really return true when its an empty object?

File: frontend/test/specs/post/postDetailsControllerSpec.js [hunk 28]
```diff
@@ -613,4 +617,71 @@
...
+            spyOn(messageService, 'showToast');
+            postDetailsCtrl.post = new Post({
+                shared_event: new Event({followers: []})
+            });
+
+            postDetailsCtrl.followEvent();
+            scope.$apply();
+
+            expect(eventService.addFollower).toHaveBeenCalled();
+            expect(messageService.showToast).toHaveBeenCalled();
```
Comment (JuliePessoa) [inline frontend/test/specs/post/postDetailsControllerSpec.js]: verify parameters

File: frontend/test/specs/post/postDetailsControllerSpec.js [hunk 29]
```diff
@@ -613,4 +617,71 @@
...
+            spyOn(messageService, 'showToast');
+            postDetailsCtrl.post = new Post({
+                shared_event: new Event({ followers: [postDetailsCtrl.user.key] })
+            });
+
+            postDetailsCtrl.unFollowEvent();
+            scope.$apply();
+
+            expect(eventService.removeFollower).toHaveBeenCalled();
+            expect(messageService.showToast).toHaveBeenCalled();
```
Comment (JuliePessoa) [inline frontend/test/specs/post/postDetailsControllerSpec.js]: verify parameters

File: frontend/post/postDetailsDirective.js (hunk 20 above)
Comment (raonismaneoto) [reply frontend/post/postDetailsDirective.js]: I would have to instantiate the event first, do u think it's worth it?

File: frontend/post/postDetailsDirective.js (hunk 21 above)
Comment (raonismaneoto) [reply frontend/post/postDetailsDirective.js]: I would have to instantiate the event first, do u think it's worth it?

File: frontend/test/specs/event/eventSpec.js (hunk 26 above)
Comment (raonismaneoto) [reply frontend/test/specs/event/eventSpec.js]: but there is only one test...

File: frontend/test/specs/post/postDetailsControllerSpec.js (hunk 27 above)
Comment (raonismaneoto) [reply frontend/test/specs/post/postDetailsControllerSpec.js]: yep, if the field shared_event is not undefined it means it is a shared_event

File: frontend/post/postDetailsDirective.js (hunk 20 above)
Comment (JuliePessoa) [reply frontend/post/postDetailsDirective.js]: it should have been instantiated already

File: frontend/post/postDetailsDirective.js (hunk 21 above)
Comment (JuliePessoa) [reply frontend/post/postDetailsDirective.js]: it should have been instantiated already

File: frontend/post/postDetailsDirective.js [hunk 30]
```diff
@@ -275,6 +275,32 @@
...
+        };
+
+        postDetailsCtrl.followEvent = () => {
+            EventService.addFollower(postDetailsCtrl.post.shared_event.key).then(() => {
+                postDetailsCtrl.post.shared_event.followers.push(postDetailsCtrl.user.key);
+                MessageService.showToast('Você receberá as atualizações desse evento.');
+            });
+        };
+
+        postDetailsCtrl.unFollowEvent = () => {
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: jsdoc

File: frontend/post/postDetailsDirective.js [hunk 31]
```diff
@@ -275,6 +275,32 @@
...
+        postDetailsCtrl.isSharedEvent = () => {
+            return postDetailsCtrl.post.shared_event; 
+        };
+
+        postDetailsCtrl.isFollowingEvent = () => {
+            const eventFollowers = (postDetailsCtrl.post.shared_event && postDetailsCtrl.post.shared_event.followers) || [];
+            return eventFollowers.includes(postDetailsCtrl.user.key);
+        };
+
+        postDetailsCtrl.followEvent = () => {
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: jsdoc

File: frontend/post/postDetailsDirective.js [hunk 32]
```diff
@@ -275,6 +275,32 @@
...
         };
 
+        /**
+         * Checks if the post is actually an event that has been shared.
+         */
+        postDetailsCtrl.isSharedEvent = () => {
+            return postDetailsCtrl.post.shared_event; 
+        };
+
+        postDetailsCtrl.isFollowingEvent = () => {
```
Comment (JuliePessoa) [inline frontend/post/postDetailsDirective.js]: jsdoc

File: frontend/event/eventDetailsDirective.js [hunk 33]
```diff
@@ -148,6 +151,33 @@
...
 
+        /**
+         * Checks if the user is following the event.
+         */
+        eventCtrl.isFollower = () => {
+            return eventCtrl.event && eventCtrl.event.followers.includes(eventCtrl.user.key);
+        };
+
+        /**
+         * Add the user as a event's follower
```
Comment (JuliePessoa) [inline frontend/event/eventDetailsDirective.js]: as an event *

**Pull request merged**